File: src/Niche/Nim/builder.nim
type
  ReportFormat = enum text, html
  ReportBuilder = object
    format: ReportFormat
    parts: seq[string]

proc reset(builder: var ReportBuilder) = builder.parts.setLen(0)

proc addTitle(builder: var ReportBuilder, title: string) =
  if builder.format == text: builder.parts.add("# " & title)
  else: builder.parts.add("<h1>" & title & "</h1>")

proc addSection(builder: var ReportBuilder, heading, body: string) =
  if builder.format == text:
    builder.parts.add("## " & heading)
    builder.parts.add(body)
  else:
    builder.parts.add("<h2>" & heading & "</h2>")
    builder.parts.add("<p>" & body & "</p>")

proc build(builder: ReportBuilder): string =
  if builder.format == text: builder.parts.join("\n")
  else: builder.parts.join("")

proc buildAvailabilityReport(builder: var ReportBuilder): string =
  builder.reset()
  builder.addTitle("Service status")
  builder.addSection("Availability", "99.95%")
  builder.build()

import std/strutils
var textBuilder = ReportBuilder(format: text)
var htmlBuilder = ReportBuilder(format: html)
echo buildAvailabilityReport(textBuilder)
echo "---"
echo buildAvailabilityReport(htmlBuilder)

File: src/Other/MicroPython/builder.py
class ReportBuilder:
    def __init__(self, format_name):
        self.format_name = format_name
        self.parts = []

    def reset(self):
        self.parts = []

    def add_title(self, title):
        if self.format_name == "text":
            self.parts.append("# " + title)
        else:
            self.parts.append("<h1>" + title + "</h1>")

    def add_section(self, heading, body):
        if self.format_name == "text":
            self.parts.append("## " + heading)
            self.parts.append(body)
        else:
            self.parts.append("<h2>" + heading + "</h2><p>" + body + "</p>")

    def build(self):
        return "\n".join(self.parts)


def build_availability_report(builder):
    builder.reset()
    builder.add_title("Service status")
    builder.add_section("Availability", "99.95%")
    return builder.build()


print(build_availability_report(ReportBuilder("text")))
print("---")
print(build_availability_report(ReportBuilder("html")))
